Initial event-native documentation (#935)
gravitee-io/gravitee-docs: pages/apim/3.x/v4-beta/v4-beta-event-native-apim-introduction.adoc

With the rise of cloud computing, many vendors started to build their solutions for the cloud. Rather than taking on-premises software and converting it to work as SaaS, it was built from the ground to to be cloud-based. These solutions are said to be cloud-native.

Event-native API management is similar. Rather than taking a synchronous API management solution and converting it to handle event-driven APIs, an event-native API management solution is built from the ground up to handle event-driven APIs.

Gravitee's API management solution can run in event-native mode, and it can support all asynchronous, streaming, and event-driven API use cases. Of course, Gravitee's API management solution also supports synchronous use cases.

Gravitee can act as a bridge between synchronous and asynchronous environments.

For example, Gravitee can subscribe to a Kafka topic and then allow filtered HTTP GET subscriptions from clients as shown in the following diagram.

image:{% link /images/apim/3.x/event-native/event-native-api-management-gateway.png %}[]

== BETA Features

The following features are available in the BETA release.

=== HTTP POST and HTTP GET entrypoints

> The HTTP POST and HTTP GET entrypoints are Community Edition / OSS plugins that enable external clients to publish and receive data via standard HTTP POST / HTTP GET requests, and ease the ingestion of data into event-oriented backends like Kafka.

Reviewer: ```suggestion
The HTTP POST and HTTP GET entrypoints are Community Edition (OSS) plugins that enable external clients to publish and receive data via standard HTTP POST / HTTP GET requests, and ease the ingestion of data into event-oriented backends like Kafka.

They are available as separate plugins from the link:https://download.gravitee.io/#graviteeio-apim/plugins/entrypoints/[entrypoints plugin download page].
```

With the HTTP POST workflow, a client application running on HTTP can post content to the Gravitee Gateway. The Gateway can then take that content and push it over to Kafka as a message that Kafka can ingest. The Gravitee Gateway acts as a protocol mediation layer in this context. This enables API owners to quickly set up a sync-to-async integration.

The diagram below shows the workflow for this use case:

image:{% link /images/apim/3.x/event-native/event-native-api-management-use-case-http-post.png %}[]

The HTTP POST entrypoint enables you to push messages to any supported event broker via a "sync call" (post) - it can be used in combination with specific Asynchronous API endpoints in different ways, depending on what publish and subscribe mode actions are possible between the two when using a connector.

> === Support for Websocket and SSE entrypoints

Reviewer: ```suggestion
=== WebSocket and SSE entrypoints
```

> Support for Server-Sent Events (Enterprise Edition only) and Websocket (Community Edition/OSS) entrypoints are provided as separate plugins, available from the link:https://download.gravitee.io/#graviteeio-apim/plugins/entrypoints/[entrypoints plugin download page]

Reviewer: ```suggestion
Server-Sent Events (SSE) and WebSocket entrypoints are available as separate plugins, available from the link:https://download.gravitee.io/#graviteeio-apim/plugins/entrypoints/[entrypoints plugin download page]. Please note that you need an Enterprise Edition (EE) license to use the SSE Advanced entrypoint capabilities. 
```

For example, if an API consumer cannot connect to a native Kafka instance due to technical limitations (for example, if they cannot implement a Kafka client themselves), Kafka topics can be made available to that consumer via a more consumer-friendly protocol such as SSE or Websocket.

The diagram below shows the workflow for this use case:

image:{% link /images/apim/3.x/event-native/event-native-api-management-use-case-event-consumption-streaming.png %}[]

This feature enables you to potentially cover use cases in the fields of Event-Driven Architecture (EDA) and Internet of Things (IoT) without the Gateway and API Management components becoming a bottleneck.

NOTE: In comparison to the link:#http_post_and_http_get_entrypoints[HTTP POST entrypoint use case] using Kafka (where the HTTP POST method is just about sending information from the client app to Kafka), with Websocket it is possible for the client app to send and receive information from Kafka as well. SSE can consume information from Kafka.

=== Webhook entrypoint

> This is a feature that enables you to front your backend with a link:https://en.wikipedia.org/wiki/Webhook[Webhook^]. It has CE and EE versions.

Reviewer: ```suggestion
This is a feature that enables you to front your backend with a link:https://en.wikipedia.org/wiki/Webhook[Webhook^]. It has CE/OSS and EE versions - Webhook, and Webhook Advanced, respectively.

Both variants are available as a plugin from the link:https://download.gravitee.io/#graviteeio-apim/plugins/entrypoints/[entrypoints plugin download page]. Please note that you need an Enterprise Edition (EE) license to use the Webhook Advanced entrypoint capabilities. 
```

= Introduction
:page-sidebar: apim_3_x_sidebar
:page-permalink: apim/3.x/event_native_apim_introduction.html
:page-folder: apim/v4-beta
:page-layout: apim3x

[label label-version]#New in version 3.20.0#
[label label-version]#BETA release#

== Event-driven asynchronous APIs

Most developers are familiar with synchronous APIs, where the client sends a request to a server, and the server sends a response back. Every time the client needs new data, it has to send a request. You can think of synchronous APIs as a 'client pull' model. RESTful APIs are a type of synchronous API.

In contrast, asynchronous APIs are event-driven - the initial request opens a connection and data (responses) is received back as and when it becomes available, where certain events would trigger the responses. Asynchronous APIs also allow you to stream data - a common use case in banking and finance systems and applications, as well as many other domains. Asynchronous APIs can use callbacks (such as synchronous API webhooks) if the requested data is not available at the time of the request. You can think of asynchronous APIs as a ‘server push’ model. AsyncAPIs are a type of asynchronous API.

> == Event-native API management

Reviewer: ```suggestion
== Gravitee's event-native API management
```

With the rise of cloud computing, many vendors started to build their solutions for the cloud. Rather than taking on-premises software and converting it to work as SaaS, it was built from the ground to to be cloud-based. These solutions are said to be cloud-native.

Event-native API management is similar. Rather than taking a synchronous API management solution and converting it to handle event-driven APIs, an event-native API management solution is built from the ground up to handle event-driven APIs.

> == Gravitee's event-native API management

Reviewer: ```suggestion
```

In contrast, asynchronous APIs are event-driven - the initial request opens a connection and data (responses) is received back as and when it becomes available, where certain events would trigger the responses. Asynchronous APIs also allow you to stream data - a common use case in banking and finance systems and applications, as well as many other domains. Asynchronous APIs can use callbacks (such as synchronous API webhooks) if the requested data is not available at the time of the request. You can think of asynchronous APIs as a ‘server push’ model. AsyncAPIs are a type of asynchronous API.

== Gravitee's event-native API management

With the rise of cloud computing, many vendors started to build their solutions for the cloud. Rather than taking on-premises software and converting it to work as SaaS, it was built from the ground to to be cloud-based. These solutions are said to be cloud-native.

> Event-native API management is similar. Rather than taking a synchronous API management solution and converting it to handle event-driven APIs, an event-native API management solution is built from the ground up to handle event-driven APIs.

Reviewer 2: This is exactly what we have done, making async and sync compatible together. I am not sure to understand the "rather" part.

Author: I understand your concerns. This paragraph uses a lot of marketing jargon to help bridge the gap between the marketing material and the technical documentation.

> Gravitee's API management solution can run in event-native mode, and it can support all asynchronous, streaming, and event-driven API use cases. Of course, Gravitee's API management solution also supports synchronous use cases.

Reviewer 2: ```suggestion
Gravitee's API management solution can run in event-native mode, and it can support all asynchronous use cases such as streaming or event-driven API. Of course, Gravitee's API management solution also supports synchronous use cases.
```

> Gravitee's API management solution can run in event-native mode, and it can support all asynchronous, streaming, and event-driven API use cases. Of course, Gravitee's API management solution also supports synchronous use cases.

Author: ```suggestion
Gravitee's API management solution can run in event-native mode, and it can support all asynchronous use cases, such as streaming or event-driven APIs. Of course, Gravitee's API management solution also supports synchronous use cases.
```

> Gravitee can act as a bridge between synchronous and asynchronous environments.

Reviewer 2: ```suggestion
Gravitee can act as a bridge between synchronous and asynchronous APIs.
```

=== Webhook entrypoint

This is a feature that enables you to front your backend with a link:https://en.wikipedia.org/wiki/Webhook[Webhook^]. It has CE/OSS and EE versions - Webhook, and Webhook Advanced, respectively. Both variants are available as part of a separate plugin you can download from the link:https://download.gravitee.io/#graviteeio-apim/plugins/entrypoints/[entrypoints plugin download page]. Please note that you need an Enterprise Edition (EE) license to use the Webhook Advanced entrypoint capabilities. 

With Webhook support, an API consumer subscribes to the Webhook API, which basically indicates to the Gateway to listen for specific Kafka messages and then call the consumer to let them know that an expected event has occurred. This communication is performed via Webhook, however it is always up to the Gateway to push the events or messages to the consumer.

The diagram below shows the workflow for this use case:

image:{% link /images/apim/3.x/event-native/event-native-api-management-use-case-event-consumption-webhook.png %}[]

> === Security policies

Reviewer 3: Previous sections were talking about the different entrypoints supported and then it switches to "Security policies" which appears not really related.

Author: These are all features that are under the heading "BETA features". Arguably, the security policies are also beta features, but because they are the same as in previous releases, I will move this heading 'up' one level. This will make it a bit clearer.

Reviewer 3: Ok but I'm still not convinced sorry. Don't you think that instead of talking about security policies we should more talk about plans ? Plan are the way to secure apis. Under the hood, plan are relying on security policy implementations but this is a technical point of view. When a user wants to secure his api, he creates a plan and select a security type (which will technically use the corresponding policy) but he never put a "security policy" by himself.

Author: OK. I think I understand now. Would it make more sense to remove this section entirely? It seems that it works the same way as in previous versions anyway.

Author: OK, how does this sound?

== Security

Gravitee can still provide the same security as in previous versions. You can secure access using an API key, JWT, or OAuth.

> === Security policies

Author: ```suggestion
== Security policies
```
